Making globals part of PSBT struct. Closes #652

dr-orlovsky opened the pull request:
I took the most non-invasive approach to reduce diff size. Many parts of the code can be improved in style or further refactored (like some functions are not necessary and can be just moved to be part of other functions), but I'd prefer to do that as a separate PR once this will be merged.

My approach with this PR:
1. Remove `Global` struct by moving its fields right into `PartiallySignedTransaction` - but keep the `util/psbt/map/global.rs` file with all its logic
2. Keep existing `Map for Global` implementation in the same file, but just change it to `Map for PartiallySignedTransaction` 
3. With serialization, convert `Global` deserialization into crate-private function and use it from `PartiallySignedTransaction` deserialization
4. Refactor the tests and imports as required to get the thing compile and pass tests

The refactoring will be followed by PR(s) adding support for Taproot

sanket1729 reviewed (commented):
cr-ACK ac0c9087dead86bc5fccdcc49c527f36578e6920. 

I don't like it if we have to call `verify()` type functions after creating an object. I guess the user can edit the unsigned tx.

dr-orlovsky commented:
@sanket1729 addressed your comment. As for the user ability to modify unsigned tx, this is a separate problem which was present before this PR. I think it should be addresses in some other PR (since there might be multiple ways of doing that). I think that we need to make the field private and provide a read-only getter for it.

apoelstra commented:
Agreed that the unsigned tx should be private and accessed through a getter. In PSBT2 there will not even necessarily be an unsigned tx, and we may need to compute it on the fly and/or cache it.

Agreed also that this change belongs in another PR.

apoelstra reviewed (approved):
ACK 55c627715f52f1377b60378241422931130583b4

I kinda think this stuff should live in psbt/mod.rs rather than psbt/global.rs, but I don't feel strongly about it.

dr-orlovsky commented:
@apoelstra I do agree that the stuff belongs to `mod`, I just wanted to reduce diff as much as possible and simplift revieweing of the actual code changes line by line, not in large blocks moved between files. We will get rid of `global.rs` as a part of further PSBT refactoring and/or splitting of this crate into a multiple crates

dr-orlovsky commented:
@sanket1729 pls re-review this PR, such that we can go forward with it and merge

sanket1729 reviewed (approved):
ACK 55c627715f52f1377b60378241422931130583b4 . Reviewed range diff with ac0c908 that I previously ACKed

JeremyRubin commented:
is there a reason this PR deletes the PSBT merge function?

dr-orlovsky commented:
It does not, it just now uses merge function which is a part of `Map` trait.

apoelstra commented:
@dr-orlovsky there used to be a publicly exposed `Psbt::merge` function which implemented the Combiner role. It looks like, since this PR, there is no such function.

We should restore it -- though I'd prefer it be called `combine` to match the PSBT naming.

tcharding commented:
FTR fixed in https://github.com/rust-bitcoin/rust-bitcoin/pull/842
